From d72fa7a842405592a3b38282e59639843147601b Mon Sep 17 00:00:00 2001
From: Steve Lhomme <slhomme@matroska.org>
Date: Fri, 10 Oct 2025 12:01:35 +0200
Subject: [PATCH] detect mkostemp with stdlib.h

On Apple platforms it may be detected as available by
just linking with it.
But when calling it it may be turned into a weak
reference that will fail to run on older platforms.

It's possible to get around this using __builtin_available() [^1]
but due to the way this code is used a clean fallback is not
possible.

It's available since macOS 10.12, iOS 10.0, tvOS 10.0 and watchOS 3.0.

[^1]: https://clang.llvm.org/docs/LanguageExtensions.html#objective-c-available
---
 configure.ac | 4 +++-
 meson.build  | 4 ++--
 2 files changed, 5 insertions(+), 3 deletions(-)

diff --git a/configure.ac b/configure.ac
index f2cf27ef..1cc95dd8 100644
--- a/configure.ac
+++ b/configure.ac
@@ -194,7 +194,9 @@ AC_TYPE_PID_T
 # Checks for library functions.
 AC_FUNC_VPRINTF
 AC_FUNC_MMAP
-AC_CHECK_FUNCS([link mkstemp mkostemp _mktemp_s mkdtemp getopt getopt_long getprogname getexecname rand random lrand48 random_r rand_r readlink fstatvfs fstatfs lstat strerror strerror_r])
+AC_CHECK_FUNCS([link mkstemp _mktemp_s mkdtemp getopt getopt_long getprogname getexecname rand random lrand48 random_r rand_r readlink fstatvfs fstatfs lstat strerror strerror_r])
+
+AC_CHECK_DECL([mkostemp],[AC_DEFINE_UNQUOTED([HAVE_MKOSTEMP],[1],[Define to 1 if you have the 'mkostemp' function.])],[],[#include <stdlib.h>])
 
 AC_SEARCH_LIBS([fabs], [m], [], [AC_MSG_ERROR([unable to find the fabs() function])])
 
diff --git a/meson.build b/meson.build
index d71643db..4202eb5f 100644
--- a/meson.build
+++ b/meson.build
@@ -118,7 +118,6 @@ check_headers = [
 check_funcs = [
   ['link'],
   ['mkstemp'],
-  ['mkostemp'],
   ['_mktemp_s'],
   ['mkdtemp'],
   ['getopt'],
@@ -156,7 +155,8 @@ check_freetype_funcs = [
 ]
 
 check_header_symbols = [
-  ['posix_fadvise', 'fcntl.h']
+  ['posix_fadvise', 'fcntl.h'],
+  ['mkostemp', 'stdlib.h'],
 ]
 
 check_struct_members = [
-- 
2.50.1 (Apple Git-155)

